modules/gapi/src/compiler/gasync.cpp

```diff
@@ -0,0 +1,133 @@
+                {
+                    //thread has not been started yet, so start it
+                    thrd = std::thread {[this](){
+                        auto second_q {q};
+                        while (!exiting){
+                            std::unique_lock<std::mutex> lck{mtx};
+                            if (q.empty())
+                            {
+                                cv.wait(lck, [&](){ return exiting || !q.empty();});
+                            }
+
+                            q.swap(second_q);
+                            lck.unlock();
+
+                            if (!second_q.empty())
+                            {
+                                auto f = second_q.front();
+                                second_q.pop();
+                                f();
+                            }
+                        }
+                    }};
+
+                }
+                std::unique_lock<std::mutex> lck{mtx};
+                bool first_task = q.empty();
+                q.push(std::move(t));
+                lck.unlock();
+                if (first_task)
+                {
+                    cv.notify_one();
+                }
+            }
+            ~async_service(){
+                exiting = true;
+                mtx.lock();
+                mtx.unlock();
+                cv.notify_one();
+                thrd.join();
+            }
+        };
+
+        async_service the_ctx;
+    }
+
+    void async_apply(GComputation& gcomp, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs, GCompileArgs &&args){
+        auto l = [=]() mutable {
+            gcomp.apply(std::move(ins), std::move(outs), std::move(args));
+            callback(std::exception_ptr {});
+        };
+        impl::the_ctx.add_task(l);
+    }
+
+    std::future<void> async_apply(GComputation& gcomp, GRunArgs &&ins, GRunArgsP &&outs, GCompileArgs &&args){
+        move_through_copy<std::promise<void>> prms{{}};
+        auto f = prms.value.get_future();
+        auto l = [=]() mutable {
+            gcomp.apply(std::move(ins), std::move(outs), std::move(args));
+            prms.value.set_value();
+        };
+
+        impl::the_ctx.add_task(l);
+        return f;
+
+    }
+
+    void async(GCompiled& gcmpld, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs){
+        auto l = [=]() mutable {
+            gcmpld(std::move(ins), std::move(outs));
+            callback(std::exception_ptr {});
+        };
+        impl::the_ctx.add_task(l);
+    }
+
+    std::future<void> async(GCompiled& gcmpld, GRunArgs &&ins, GRunArgsP &&outs){
+        move_through_copy<std::promise<void>> prms{{}};
+        auto f = prms.value.get_future();
+        auto l = [=]() mutable {
+            gcmpld(std::move(ins), std::move(outs));
+            prms.value.set_value();
+        };
+
+        impl::the_ctx.add_task(l);
+        return f;
+    }
+}
```

Comment:
Is there anywhere any info describing this?

Reply:
(seems I got it after reading, but anyway overall description that "now it is just a thread with a queue, calling gcompileds followed by their completion callbacks" is kinda missing

Reply:
added